File: backJoon/src/Q3273_두_수의_합/main2.cpp
//
// Created by 김종현 on 2023/05/11.
//
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int N, cmp, ans;
vector<int> v;

void input() {
    ios::sync_with_stdio(0), cin.tie(0), cout.tie(0);
    cin >> N;
    int num;
    for (int i = 0; i < N; i++) {
        cin >> num;
        v.push_back(num);
    }
    sort(v.begin(), v.end());
    cin >> cmp;
}

void solve() {
    int l = 0;
    int r = N - 1;
    while (l < r) {
        int num = v[l] + v[r];
        if (num == cmp) l++, r--, ans++;
        else if (num < cmp) l++;
        else r--;
    }
}

void output() {
    cout << ans;
}

int main() {
    input();
    solve();
    output();
}

File: backJoon/src/Q14950_정복자/Main2.java
package Q14950_정복자;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main2 {
    int N, M, T, edges[][], p[], ans;

    private void input() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer stk = new StringTokenizer(br.readLine());

        N = Integer.parseInt(stk.nextToken());
        M = Integer.parseInt(stk.nextToken());
        T = Integer.parseInt(stk.nextToken());

        edges = new int[M][3];

        p = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            p[i] = i;
        }

//        edges[0] = new int[]{0, 1, 0};
        for (int i = 0; i < M; i++) {
            stk = new StringTokenizer(br.readLine());
            edges[i] = new int[]{Integer.parseInt(stk.nextToken()), Integer.parseInt(stk.nextToken()), Integer.parseInt(stk.nextToken())};
        }

        Arrays.sort(edges, (n1, n2) -> n1[2] - n2[2]);
    }

    private int find(int num) {
        if (p[num] == num) return num;
        return p[num] = find(p[num]);
    }

    private void solve() {
        int cnt = 0;
        for (int i = 0; i < M; i++) {
            int n1 = find(edges[i][0]);
            int n2 = find(edges[i][1]);

            if (n1 != n2) {
                p[n1 < n2 ? n2 : n1] = n1 < n2 ? n1 : n2;
                ans += edges[i][2] + T * cnt++;
            }
            if (cnt == N) break;
        }
    }

    private void output() {
        System.out.println(ans);
    }

    public static void main(String[] args) throws Exception {
        Main2 m = new Main2();
        m.input();
        m.solve();
        m.output();
    }
}
